src/selectors/disabled.test.js

```diff
@@ -0,0 +1,164 @@
+  getNodeDisabled,
+  getEdgeDisabled
+} from './disabled';
+import { toggleNodesDisabled } from '../actions/nodes';
+import { toggleTagFilter } from '../actions/tags';
+import reducer from '../reducers';
+
+const getNodeIDs = state => state.node.ids;
+const getEdgeIDs = state => state.edge.ids;
+const getEdgeSources = state => state.edge.sources;
+const getEdgeTargets = state => state.edge.targets;
+const getNodeTags = state => state.node.tags;
+const getNodeType = state => state.node.type;
+
+describe('Selectors', () => {
+  describe('getNodeDisabledTag', () => {
+    it('returns an object', () => {
+      expect(getNodeDisabledTag(mockState.lorem)).toEqual(expect.any(Object));
+    });
+
+    it("returns an object whose keys match the current pipeline's nodes", () => {
+      expect(Object.keys(getNodeDisabledTag(mockState.lorem))).toEqual(
+        getNodeIDs(mockState.lorem)
+      );
+    });
+
+    it('returns an object whose values are all Booleans', () => {
+      expect(
+        Object.values(getNodeDisabledTag(mockState.lorem)).every(
+          value => typeof value === 'boolean'
+        )
+      ).toBe(true);
+    });
+
+    it('does not disable a node if all tags are disabled', () => {
+      const nodeDisabled = getNodeDisabledTag(mockState.lorem);
+      expect(Object.values(nodeDisabled)).toEqual(
+        Object.values(nodeDisabled).map(() => false)
+      );
+    });
+
+    it('disables a node only if all of its tags are disabled', () => {
+      const nodeTags = getNodeTags(mockState.animals);
+      // Get list of task nodes from the current pipeline
+      const taskNodes = getNodeIDs(mockState.animals).filter(
+        id => getNodeType(mockState.animals)[id] === 'task'
+      );
+      // Choose a node that has some tags (and which should be enabled)
+      const hasTags = id => Boolean(nodeTags[id].length);
+      const enabledNodeID = taskNodes.find(hasTags);
+      // Choose a node that has no tags (and which should be disabled)
+      const hasNoTags = id => !Boolean(nodeTags[id].length);
+      const disabledNodeID = taskNodes.find(hasNoTags);
+      // Update the state to enable one of the tags for that node
+      const enabledNodeTags = nodeTags[enabledNodeID];
+      const newMockState = reducer(
+        mockState.animals,
+        toggleTagFilter(enabledNodeTags[0], true)
+      );
+      expect(getNodeDisabledTag(newMockState)[enabledNodeID]).toEqual(false);
+      expect(getNodeDisabledTag(newMockState)[disabledNodeID]).toEqual(true);
+    });
+  });
+
+  describe('getNodeDisabled', () => {
+    it('returns an object', () => {
+      expect(getNodeDisabled(mockState.lorem)).toEqual(expect.any(Object));
+    });
+
+    it("returns an object whose keys match the current pipeline's nodes", () => {
+      expect(Object.keys(getNodeDisabled(mockState.lorem))).toEqual(
+        getNodeIDs(mockState.lorem)
+      );
+    });
+
+    it('returns an object whose values are all Booleans', () => {
```

Comment:
I think the following tests (`'does not disable a node if all tags are disabled'` and `'disables a node only if all of its tags are disabled'`) accomplish that goal? However they do make this test is a bit redundant. I just added it so that if it fails then it's clear that something's very wrong with the output format, not just with the logic. Do you reckon I should remove it as it's redundant?

In this case, `nodeDisabledTag` returns an object containing nodeID keys and boolean values, which show whether a node has been disabled due to **all** of its tags being disabled, if a tag filter is active. They should all be false by default. I might add some more specific tests.

Worth noting that I just moved this selector/test code from one file to another during this PR - I didn't write it.

Reply:
Looks like that is covered then, think I got a little confused when I read it. May be a little redundant yeah but no problem to keep it in I think.